Neuron observability

freschri opened the pull request:
Single New EKS Cluster Open Source Observability Accelerator for Neuron-based clusters
depends on https://github.com/aws-observability/aws-observability-accelerator/pull/32
@elamaran11 please check if we want to keep/remove cdk.json from gitignore (we need to be able to add entries that are validated in compilation otherwise github compilation fails on push)

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+# Single New EKS Cluster Open Source Observability Accelerator for Neuron-based clusters
```
Rename this to `Single Cluster Open Source Observability - Neuron-based cluster`

freschri replied:
done

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+
+[AWS Neuron](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/) is an SDK with a compiler, runtime, and profiling tools that helps developers deploy models on both AWS Inferentia accelerators and train them on AWS Trainium accelerators. It integrates natively with popular ML frameworks, such as PyTorch and TensorFlow.
+
+This pattern shows you how to monitor the performance of ML accelerators, used in an Amazon EKS cluster leveraging Inferentia-based Amazon EC2 Inf1 instances. The pattern will also be applicable to Inf2 and Trn1/Trn1n instances soon, when AWS Cloud Development Kit (CDK) completes support for Inf2 and Trn1/Trn1n (in progress, at the time of writing).
```
rename accelerators to workloads

freschri replied:
that's their name.. workloads is not right in this context. e.g. see here https://aws.amazon.com/machine-learning/neuron/ "AWS Trainium and AWS Inferentia accelerators "

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+
+The AMI type of the node group is `AL2_x86_64_GPU AMI`, which uses the [Amazon EKS-optimized accelerated AMI](https://aws.amazon.com/marketplace/pp/prodview-nwwwodawoxndm). In addition to the standard Amazon EKS-optimized AMI configuration, the accelerated AMI includes the AWS Neuron container runtime.
+
+The Neuron container runtime consists of kernel driver and C/C++ libraries which provide APIs to access Inferentia and Trainium Neuron devices. The Neuron ML frameworks plugins for TensorFlow, PyTorch and Apache MXNet use the Neuron runtime to load and run models on the NeuronCores.
```
NeuronCores -> Neuron Cores

freschri replied:
that's the name, please see here: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-hardware/neuroncores-arch.html

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+
+`neuron-monitor` collects metrics and stats from the Neuron Applications running on the system and streams the collected data to stdout [in JSON format](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/tools/neuron-sys-tools/neuron-monitor-user-guide.html#the-json-output-format).
+
+`neuron-monitor-prometheus.py` maps and exposes the telemetry data from JSON format into Prometheus compatible format.
```
where is `neuron-monitor-prometheus.py` ?

freschri replied:
it is explained a few lines above, "Specifically, the container runs the neuron-monitor command piped into the `neuron-monitor-prometheus.py` companion script", so it is inside the container

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+        "GRAFANA_NODEEXP_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodeexporter-nodes.json",
+        "GRAFANA_NODES_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodes.json",
+        "GRAFANA_WORKLOADS_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/workloads.json",
+        "GRAFANA_NEURON_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/neuron/neuron-monitor.json"
```
Dont you need to add this dash to `aws-observability-accelerato`r repo?

freschri replied:
yes I mentioned the dependency in the description of this pr: "depends on https://github.com/aws-observability/aws-observability-accelerator/pull/32"

elamaran11 commented on lib/single-new-eks-opensource-observability-pattern/neuron/neuron-device-plugin-addon.ts:
```diff
@@ -0,0 +1,44 @@
+import { KubernetesManifest, ServiceAccount } from "aws-cdk-lib/aws-eks";
```
You should move this to blueprints repo. I see an active PR - https://github.com/aws-quickstart/cdk-eks-blueprints/pull/777 for this. Lets get this from blueprints vs using it here. Having in blueprints make this available to all.

elamaran11 commented on lib/single-new-eks-opensource-observability-pattern/neuron/neuron-device-plugin-role.yaml:
```diff
@@ -0,0 +1,32 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
```
this should also move to blueprints repo.

elamaran11 commented on lib/single-new-eks-opensource-observability-pattern/neuron/neuron-monitor-addon.ts:
```diff
@@ -0,0 +1,18 @@
+import { KubernetesManifest } from "aws-cdk-lib/aws-eks";
```
This should also move to blueprints

freschri replied:
done

elamaran11 commented on lib/single-new-eks-opensource-observability-pattern/neuron/neuron-monitor.yaml:
```diff
@@ -0,0 +1,85 @@
+kind: DaemonSet
```
This should also move to blueprints

freschri replied:
done

elamaran11 commented on lib/common/resources/neuron/pytorch-inference-resnet50.yml:
```diff
@@ -0,0 +1,96 @@
+apiVersion: v1
```
This file should be moved to aws-observability-accelerator-repo and i would prefer using a GitOps approach to deploy the workload

freschri replied:
it is an example workload, just generating metrics, not a real one never to be used in production.
so I believe deployed manually as a pod is better?

elamaran11 replied:
I would say deploy this manually as a pattern. DockerFile should be moved to a different repo may be `aws-observability-accelerator-repo `

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
```diff
@@ -0,0 +1,268 @@
+
+Amazon EC2 ML instances belong to the [Trn1/Trn1n](https://aws.amazon.com/ec2/instance-types/trn1/), [Inf1](https://aws.amazon.com/ec2/instance-types/inf1/) and [Inf2](https://aws.amazon.com/ec2/instance-types/inf2/) families. Trn1/Trn1n instances feature multiple AWS Trainium accelerators and support high-performance training. Inf1 and Inf2 instances feature multiple AWS Inferentia accelerators  ...
+
+[AWS Neuron](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/) is an SDK with a compiler, runtime, and profiling tools that helps developers deploy models on both AWS Inferentia accelerators and train them on AWS Trainium accelerators. It integrates natively with popular ML frameworks, such as PyTorch and TensorFlow.
```
There term accelerators is unclear here

freschri replied:
it is explained at the beginning of the doc: "[AWS Trainium]and [AWS Inferentia] are accelerated Machine Learning (ML) chips (or ML accelerators)", see also here: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/glossary.html?highlight=glossary#neuron-devices-accelerated-machine-learning-chips

elamaran11 commented on docs/patterns/single-new-eks-observability-accelerators/single-new-eks-inferentia-neuron-opensource-observability.md:
````diff
@@ -0,0 +1,268 @@
+    --region us-east-2
+```
+
+8. For the `neuron-monitor` DaemonSet, you can either use the image already referenced into the manifest, or build your own using the Dockerfile at location `cdk-aws-observability-accelerator/lib/common/resources/neuron/neuron-monitor.dockerfile`, [push it to an ECR repository](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html) of your choice and update the image U ...
````
`neuron-monitor` should be moved to blueprints addon. You can move the `neuron-monitor.dockerfile` to `aws-observability-accelerator` repo and you can show in blueprints documentation on how to build the docker file with addon document.

freschri replied:
done

elamaran11 reviewed (changes requested):
@freschri This is AWSome work. We need some strategic and tactical changes as per the comments below to align this well. Love the neuron pattern.

elamaran11 commented:
@freschri Any updates to this PR?

freschri commented:
> @freschri Any updates to this PR?

I addressed all points apart from the addons. One of them depends on another PR in Blueprints (Neuron Device Plugin Addon). My suggestion was on another channel to merge this PR as is, if no other concern, and raise 2 issues to port the addons to Blueprints. That can be done as an exercise by someone with limited experience on Blueprints/Patterns/Observability. Please let me know

elamaran11 commented:
> > @freschri Any updates to this PR?
> 
> I addressed all points apart from the addons. One of them depends on another PR in Blueprints (Neuron Device Plugin Addon). My suggestion was on another channel to merge this PR as is, if no other concern, and raise 2 issues to port the addons to Blueprints. That can be done as an exercise by someone with limited experience on Blueprints/Patterns/Observability. Please let me know

Thats a good opportunity mentor, if you have some one in your region. If i have someone, i will let you know. But this is not a burning one, please take it slow.

ariveroi commented:
New pr against inf1 branch with new quickstart version: https://github.com/aws-observability/cdk-aws-observability-accelerator/pull/153

If we merge #153 PR to inf1 branch we will be able to merge this PR

elamaran11 commented on lib/common/resources/otel-collector-config-new.yml:
```diff
@@ -0,0 +1,1622 @@
+#
```
Remove this file, should not be checked in!

freschri replied:
done

elamaran11 commented on lib/common/resources/neuron/pytorch-inference-resnet50.yml:
```diff
@@ -0,0 +1,96 @@
+apiVersion: v1
+kind: ConfigMap
```
Move this file outside of repo!

freschri replied:
would this be a good location: https://github.com/aws-observability/aws-observability-accelerator/tree/main/artifacts/k8s-deployment-manifest-templates ?

elamaran11 replied:
Works for me!

elamaran11 replied:
create a new folder like `neuron`

freschri replied:
done, see pr https://github.com/aws-observability/aws-observability-accelerator/pull/39

elamaran11 commented on .gitignore:
```diff
@@ -12,7 +12,6 @@ node_modules
 # CDK asset staging directory
 .cdk.staging
 cdk.out
-cdk.json
```
Why is this removed?

freschri replied:
done

elamaran11 commented on bin/single-new-eks-inferentia-opensource-observability.ts:
```diff
@@ -0,0 +1,6 @@
+
+const app = configureApp();
+
+new SingleNewEksInferentiaOpenSourceObservabilityPattern(app, 'single-new-eks-inferentia-opensource');
```
Update the `readMe.md` `make list` command output with this new pattern.

freschri replied:
it is there

elamaran11 commented on cdk.json:
```diff
@@ -32,21 +32,33 @@
         "GRAFANA_NODES_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/nodes.json",
-        "GRAFANA_WORKLOADS_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/workloads.json"
+        "GRAFANA_WORKLOADS_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/infrastructure/workloads.json",
+        "GRAFANA_NEURON_DASH_URL" : "https://raw.githubusercontent.com/aws-observability/aws-observability-accelerator/main/artifacts/grafana-dashboards/eks/neuron/neuron-monitor.json"
```
Remove this from `cdk.json` this should go in to doc only. We dont add the extra boards for all patterns.

freschri replied:
done

elamaran11 commented on lib/single-new-eks-fargate-opensource-observability-pattern/index.ts:
```diff
@@ -122,8 +122,7 @@ export default class SingleNewEksFargateOpenSourceObservabilityConstruct {
             new blueprints.addons.VpcCniAddOn(),
-            new blueprints.addons.CoreDnsAddOn({
-                version: "v1.10.1-eksbuild.6",
+            new blueprints.addons.CoreDnsAddOn("v1.10.1-eksbuild.6", {
```
version can be populated with auto or can be removed!

freschri replied:
done

elamaran11 reviewed (changes requested):
@freschri Couple of minor ones. Once you fix these, i can run e2e.

elamaran11 commented:
/do-e2e-test single-new-eks-opensource-observability deploy

elamaran11 commented:
/do-e2e-test single-new-eks-opensource-observability deploy

elamaran11 commented on package.json:
```diff
@@ -24,16 +24,16 @@
-    "@aws-quickstart/eks-blueprints": "^1.13.1",
-    "aws-cdk": "2.114.1",
+    "@aws-quickstart/eks-blueprints": "^1.14.0",
+    "aws-cdk": "2.130.0",
```
"aws-cdk-lib": "2.132.0",
        "aws-cdk": "2.132.0"

elamaran11 reviewed (changes requested):
Update the CDK version.

```
        "aws-cdk-lib": "2.132.0",
        "aws-cdk": "2.132.0"
```

elamaran11 commented:
/do-e2e-test single-new-eks-inferentia-opensource-observability deploy

elamaran11 commented:
/do-e2e-test single-new-eks-opensource-observability deploy

elamaran11 commented on cdk.json:
```diff
@@ -37,17 +37,28 @@
       "kustomizations": [
         {
           "kustomizationPath": "./artifacts/grafana-operator-manifests/eks/infrastructure"
+        },
```
Remove all neuron references from `cdk.json`. `cdk.json` is placeholder for generic pattern.

elamaran11 reviewed (changes requested):
One more feedback

freschri commented:
@elamaran11 updated blueprints version by merging main into here. please trigger e2e thanks

elamaran11 commented:
/do-e2e-test single-new-eks-inferentia-opensource-observability deploy

freschri commented:
@elamaran11 e2e passes. shall we merge?

elamaran11 commented:
> @elamaran11 e2e passes. shall we merge?

Nope lets wait, I might to crash e2e and do few other checks.

elamaran11 commented:
The E2E Failure is related to subnet ENI which is very interim. It worked on retry,

elamaran11 commented:
/do-e2e-test single-new-eks-opensource-observability deploy

elamaran11 commented:
/do-e2e-test single-new-eks-opensource-observability destroy

elamaran11 commented:
/do-e2e-test single-new-eks-inferentia-opensource-observability deploy

elamaran11 commented:
/do-e2e-test single-new-eks-inferentia-opensource-observability destroy

elamaran11 reviewed (approved):
LGTM
